### ISSUE TITLE
Dependency conflict resolution between platforms changed behavior in 8.8

### ISSUE THREAD

**Issue by TWiStErRob** [gradle/gradle#29393]
### Current Behavior

Gradle 8.8:

```text
$ gradlew depI --configuration compileClasspath --dependency kotlin-bom

> Task :dependencyInsight
org.jetbrains.kotlin:kotlin-bom:1.4.32 FAILED
   Failures:
      - Could not resolve org.jetbrains.kotlin:kotlin-bom:{strictly 1.4.32}.
          - Cannot find a version of 'org.jetbrains.kotlin:kotlin-bom' that satisfies the version constraints:
               Dependency path ':dependency_conflict-resolution-gradle88:unspecified' --> 'org.jetbrains.kotlin:kotlin-bom:{strictly 1.9.24}'
               Dependency path ':dependency_conflict-resolution-gradle88:unspecified' --> 'org.jetbrains.kotlin:kotlin-bom:{strictly 1.4.32}'

org.jetbrains.kotlin:kotlin-bom:{strictly 1.4.32} -> 1.4.32 FAILED
\--- compileClasspath

org.jetbrains.kotlin:kotlin-bom:1.9.24 FAILED
   Failures:
      - Could not resolve org.jetbrains.kotlin:kotlin-bom:{strictly 1.9.24}. (already reported)

org.jetbrains.kotlin:kotlin-bom:{strictly 1.9.24} -> 1.9.24 FAILED
\--- compileClasspath
```

### Expected Behavior

Gradle 8.7:

```text
$ gradlew depI --configuration compileClasspath --dependency kotlin-bom

> Task :dependencyInsight
org.jetbrains.kotlin:kotlin-bom:1.9.24
  Variant enforced-platform-compile:
    | Attribute Name                     | Provided          | Requested         |
    |------------------------------------|-------------------|-------------------|
    | org.gradle.status                  | release           |                   |
    | org.gradle.category                | enforced-platform | enforced-platform |
    | org.gradle.usage                   | java-api          | java-api          |
    | org.gradle.dependency.bundling     |                   | external          |
    | org.gradle.jvm.environment         |                   | standard-jvm      |
    | org.gradle.jvm.version             |                   | 17                |
    | org.gradle.libraryelements         |                   | classes           |
    | org.jetbrains.kotlin.platform.type |                   | jvm               |
   Selection reasons:
      - By conflict resolution: between versions 1.9.24 and 1.4.32

org.jetbrains.kotlin:kotlin-bom:{strictly 1.9.24} -> 1.9.24
\--- compileClasspath

org.jetbrains.kotlin:kotlin-bom:1.4.32 -> 1.9.24
\--- compileClasspath
```

### Context (optional)

I'm very likely doing something wrong here with this setup, as reducing it to a minimal example looks very strange. But I still wanted to report, in case there's some unintended changes from all the dependency resolution error messaging improvements in Gradle 8.8.

### Steps to Reproduce

1. Clone https://github.com/TWiStErRob/repros/tree/main/gradle/dependency_conflict-resolution-gradle88
2. `gradlew depI --configuration compileClasspath --dependency kotlin-bom`

### Gradle version

Gradle 8.8-rc-1, Gradle 8.8-rc-2, Gradle 8.8

### Gradle version that used to work

Gradle 8.7

**Comment by ljacomet** [comment]
The issue is in the backlog of the relevant team and is prioritized by them.

***

Thanks. We will take a look and confirm the regression aspect.

**Comment by octylFractal** [comment]
This change was made on purpose to fix #27282, as the version catalog behavior was inconsistent with the long-standing behavior of `enforcedPlatform` on `String`-based dependencies.

We considered if a behavior change to what the version catalogs were doing would be appropriate here, which results in a resolution like this:
```
compileClasspath - Compile classpath for null/main.
+--- org.jetbrains.kotlin:kotlin-bom:1.9.24
|    \--- org.jetbrains.kotlin:kotlin-stdlib:1.9.24 (c)
+--- org.jetbrains.kotlin:kotlin-stdlib:{strictly 1.4.32} -> 1.9.24 // This is forced
|    \--- org.jetbrains:annotations:13.0
\--- org.jetbrains.kotlin:kotlin-bom:1.4.32 -> 1.9.24 (*) // Normal version conflict handling changes this to the same version as the other bom
```
However, this has other problems. A higher-version regular dependency or `platform` with the same coordinates could win over `enforcedPlatform` itself. This will update the version of the `enforcedPlatform` and force the higher versions instead. This conflicts with the idea of `enforcedPlatform`, where you want _exactly_ the dependencies specified in it, regardless of other dependencies or platforms. We're still unsure which is the most user-friendly behavior, and there may also be other options, so if you still feel strongly that this is incorrect, please open a new issue describing your use-case and what you think the behavior should be.
